### ISSUE TITLE
ignore_changes not ignoring a change which uses a dynamic value

### ISSUE THREAD

**Issue by anuraaga** [hashicorp/terraform#22191]
### Terraform Version

```
Terraform v0.12.5
```

### Terraform Configuration Files

```hcl
resource kubernetes_deployment deployment {
  lifecycle {
    ignore_changes = [spec[0].template[0].metadata[0].labels["revision"]]
  }
  spec {
    template {
      metadata {
        labels = {
          name = var.name
          revision = "terraform-${timestamp()}"  // Key line
...
}
```

### Expected Behavior

Change is ignored because the value is changed, not added or removed.

### Actual Behavior

Change is not ignored with output like
```
                  ~ labels      = {
                      - "name"     = "server-alpha"
                      - "revision" = "8d6c139eeded8df121fa13aca5404e6bb60a2dca"
                    } -> (known after apply)
```

If the revision is set to just "terraform" in the config, it works - as you can see, the value changes from that git hash to the word "terraform" but plan shows no updates. I don't think it should matter whether the value is dynamic or not for whether it is ignored or not.

### Steps to Reproduce

1. `terraform init`
2. `terraform plan`

**Comment by teamterraform** [comment]
Hello @anuraaga. This could be a bug, in the meantime could I suggest a few workarounds. Could you try ignoring changes to the entire map.

```
ignore_changes = [spec[0].template[0].metadata[0].labels]
```

Or if that can't be done for your application, could you rely on another mechanism other than `timestamp()`, such as having a generated ID passed in as a terraform variable. We believe the issue is related to the fact that `timestamp()` is only known during the apply phase causing a very subtle bug.

**Comment by anuraaga** [comment]
Yup thanks for the hint - using timestamp was only to be a bit more precise but not required. I've already gone ahead and stuck with the constant for my project. Filed the issue so others don't run into it in the future.

**Comment by gavinreaney** [comment]
I've run into this too for timestamp() and the workarounds aren't applicable for us (maybe we can pass a timestamp on the command-line but it's annoying to pass that through every module).

**Comment by phistrom** [comment]
Still seeing this in Terraform 0.12.6.

I have a "Created On" tag in almost every resource that supports tagging in AWS, and I set the tag's value to `timestamp()`. If I specify `ignore_changes = [tags["Created On"]] `, and then do a `terraform apply`, I am shown an enormous number of changes are going to be made for these resources and their dependents. However, when I type "yes" to apply the changes, I get `Apply complete! Resources: 0 added, 0 changed, 0 destroyed.`

To easily see this in action, here is a slightly modified version of the HelloWorld example shown on the [aws_instance resource](https://www.terraform.io/docs/providers/aws/r/instance.html) page:


```hcl
// example.tf
provider "aws" { region = "us-west-2" }
data "aws_ami" "ubuntu" {
  most_recent = true
  filter {
    name   = "name"
    values = ["ubuntu/images/hvm-ssd/ubuntu-*-18.04-amd64-server-*"]
  }
  filter {
    name   = "virtualization-type"
    values = ["hvm"]
  }
  owners = ["099720109477"] # Canonical
}
resource "aws_instance" "web" {
  ami           = data.aws_ami.ubuntu.id
  instance_type = "t2.micro"
  tags = {
    Name = "HelloWorld2"
    "Created On" = timestamp()
  }
  lifecycle {
    ignore_changes = [
      tags["Created On"]  // This will forever show a change
    ]
  }
}
```

And the output is always like this:
```
...
Plan: 0 to add, 1 to change, 0 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes


Apply complete! Resources: 0 added, 0 changed, 0 destroyed.
```

This bug does not occur if you have `ignore_changes = ["tags"]`, but I don't use it because I have tags I _do_ want Terraform to enforce.
